topdown/topdown_test.go

```diff
@@ -1465,6 +1465,84 @@ func TestTopDownTime(t *testing.T) {
 
 }
 
+func TestTopDownWalkBuiltin(t *testing.T) {
+
+	tests := []struct {
+		note     string
+		rules    []string
+		expected interface{}
+	}{
+		{
+			note: "scalar",
+			rules: []string{
+				`p[x] { walk(data.a[0], x) }`,
+			},
+			expected: `[
+				[[], 1]
+			]`,
+		},
+		{
+			note: "arrays",
+			rules: []string{
+				`p[x] { walk(data.a, x) }`,
+			},
```

Comment:
Walk() is nondeterministic?  It returns a different x each time it is called in this test. If we call walk from 2 different rules with 2 different evaluation contexts but the same input, the walking of the datastructure should restart.  We don't have any builtins like that so far.  

Feels like a new kind of iterator.  We have iterators that work for the top-most level of an object: arr[k] = v; dict[k] = v; set[k].  This seems to be a new way to iterate over an object where instead of 1 level, it walks over all levels recursively.  Wonder if we need a new syntax for it.

An alternative is to have walk return the entire list of path/value pairs. 
`walk(thing, [[path1, value1],...])`
That would look more like the builtins we have today.